tsdb/wal/wal.go

```diff
@@ -757,23 +769,52 @@ func listSegments(dir string) (refs []segmentRef, err error) {
 		return nil, err
 	}
 	var last int
+	// assume segments in sequence
+	inSequence := true
 	for _, fn := range files {
 		k, err := strconv.Atoi(fn)
 		if err != nil {
 			continue
 		}
-		if len(refs) > 0 && k > last+1 {
-			return nil, errors.New("segments are not sequential")
+
+		if len(refs) > 0 && k > last+1 && inSequence {
+			inSequence = false
 		}
 		refs = append(refs, segmentRef{name: fn, index: k})
 		last = k
 	}
+	if !inSequence {
+		return refs, errors.New(ErrorUnsequentialSegments)
```

Comment:
It is quite unusual that `refs` is also returned with the error
I would just return nil here and the repair can read the segment numbers and do the repair complitely separately. It would duplicate some code, but would be easier to follow.